lib/core/src/Test/QuickCheck/Instances/UTCTime.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Provides a `Arbitrary` instance for `UTCTime` that spans a wide range of
+-- dates and a mixture of time precisions.
+
+module Test.QuickCheck.Instances.UTCTime ()
+    where
+
+import Prelude
+
+import Data.Time
+    ( UTCTime (..) )
+import Data.Time.Calendar
+    ( Day (ModifiedJulianDay), toModifiedJulianDay )
+import Data.Time.Clock
+    ( diffTimeToPicoseconds, picosecondsToDiffTime, secondsToDiffTime )
+import Test.QuickCheck
+    ( Arbitrary, arbitrary, choose, oneof, shrink )
+
+-- | Approximate number of days in one thousand years.
+daysInOneThousandYears :: Integral a => a
+daysInOneThousandYears = 365 * 1000
+
+-- | The maximum number of picoseconds in one day, allowing for leap seconds.
+picosecondsInOneDay :: Integral a => a
+picosecondsInOneDay = secondsInOneDay * picosecondsInOneSecond
+
+-- | The exact number of picoseconds in one second.
+picosecondsInOneSecond :: Integral a => a
+picosecondsInOneSecond = 1_000_000_000_000
+
+-- | The maxmimum number of seconds in one day, allowing for leap seconds.
+secondsInOneDay :: Integral a => a
+secondsInOneDay = (60 * 60 * 24) + 1
+
+-- | An `Arbitrary` instance for `UTCTime` that spans a wide range of dates
+--   and a mixture of time precisions.
+instance Arbitrary UTCTime where
+
+    arbitrary = oneof
+        -- Generate times with a mixture of precisions:
+        [ genAnyTime
+            secondsToDiffTime
+            secondsInOneDay
+        , genAnyTime
+            picosecondsToDiffTime
+            picosecondsInOneDay
+        ]
+      where
+        genAnyTime unitsToDiffTime unitsInOneDay = do
+            numberOfDays <- ModifiedJulianDay
+                <$> choose (0, daysInOneThousandYears)
+            timeSinceMidnight <- unitsToDiffTime
+                <$> choose (0, unitsInOneDay)
+            pure $ UTCTime numberOfDays timeSinceMidnight
+
+    shrink t@(UTCTime numberOfDays timeSinceMidnight) = mconcat
+        [ [ t { utctDay = ModifiedJulianDay x }
+            | x <- shrink $ toModifiedJulianDay numberOfDays ]
+        , [ t { utctDayTime = picosecondsToDiffTime y }
+            | y <- shrink $ diffTimeToPicoseconds timeSinceMidnight ]
+        ]
```

Comment:
Not sure how relevant it is to shrink an arbitrary time ... :thinking: 
I don't think it helps getting a _simpler_ value in anyway :sweat_smile: ? It justs get another arbitrary time and burns CPU cycles, no?

Reply:
> Not sure how relevant it is to shrink an arbitrary time ... thinking
> I don't think it helps getting a _simpler_ value in anyway sweat_smile ? It justs get another arbitrary time and burns CPU cycles, no?

Fair enough! My original thinking was to:

1. _Improve upon_ the _problematic_ behaviour of the original instance (generation).
2. _Preserve non-problematic behaviour_ as is (shrinking).

However, I also can't think of a useful use case for context-free shrinking. I'll remove the shrinker.